be/src/storage/persistent_index.cpp

```diff
@@ -212,21 +241,89 @@ Status PersistentIndex::create(size_t key_size, const EditVersion& version) {
     return Status::OK();
 }
 
-Status PersistentIndex::load() {
-    // TODO: load from index file
-    return Status::NotSupported("TODO");
+Status PersistentIndex::load(Tablet* tablet) {
+    RETURN_IF_ERROR(
+            TabletMetaManager::get_persistent_index_meta(tablet->data_dir(), tablet->tablet_id(), &_index_meta));
+    size_t key_size = _index_meta.key_size();
+    _size = _index_meta.size();
+    DCHECK_EQ(key_size, _key_size);
+    fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+    std::unique_ptr<fs::ReadableBlock> rblock;
+    DeferOp close_block([&rblock] { rblock->close(); });
+    RETURN_IF_ERROR(block_mgr->open_block(_path, &rblock));
+    // TODO: load snapshot first
+    if (!_index_meta.has_l0_meta()) {
+        return Status::OK();
+    }
+    MutableIndexMetaPB* l0_meta = _index_meta.mutable_l0_meta();
+    int n = l0_meta->wals_size();
+    // read wals and build l0
+    for (int i = 0; i < n; i++) {
+        auto* wal_pb = l0_meta->mutable_wals(i);
+        auto* page_pb = wal_pb->mutable_data();
+        size_t offset = page_pb->offset();
+        size_t size = page_pb->size();
+        _offset = offset + size;
+        std::string buff;
+        raw::stl_string_resize_uninitialized(&buff, size);
+        RETURN_IF_ERROR(rblock->read(offset, buff));
+
+        size_t nums = (size / (key_size + sizeof(IndexValue)));
+        uint8_t keys[key_size * nums];
```

Comment:
maybe stack overflow

Reply:
> maybe stack overflow

done, change in https://github.com/StarRocks/starrocks/pull/3399/commits/a2a599a3def12e52333bcdfc189224b62b7514be